src/components/TreebankService/TreebankService.js

```diff
@@ -0,0 +1,78 @@
+import { MessagingService, ResponseMessage, WindowIframeDestination as Destination } from 'alpheios-messaging/dist/dev/alpheios-messaging.js'
+import { TreebankDestinationConfig } from './TreebankDestinationConfig.js'
+
+class TreebankService {
+  constructor(arethusa) {
+    const messagingServiceName = 'TreebankRequestListener'
+    /**
+     * supportedRequests contains the supported Arethusa API requests in the format
+     * <requestName>: [ <requiredParameters> ]
+     */
+    const supportedRequests = {
+      getMorph: ['sentenceId','wordId'],
+      refreshView: [],
+      nextSentence: [],
+      prevSentence: [],
+      gotoSentence: ['sentenceId']
+    }
+    this.activate = this.activate.bind(this)
+    this.arethusa = arethusa
+
+    /**
+     * alpheios-messaging MessageService message handler function
+     * @param {Object} request is a request object. It has one property 'body', an object
+     *                 which has as it's only property named per the service request to execute
+     *                 which is itself an object with the service method's parameters and their 
+     *                 values
+     * @param {Function} responseFn a callback function which sends the service response to the client
+     */
+    this.messageHandler = (request, responseFn) => {
+      let requestName
+      for (let name in supportedRequests) {
+        if (request.body[name]) {
+          requestName = name
+          break
+        }
+      }
+      if (! requestName) {
+        responseFn(ResponseMessage.Error(request, new Error(`Unsupported request. Valid requests are ${Object.keys.supportedRequests}`)))
+        return
+      }
+      let widget = this.arethusa.getWidget()
+      let api = widget.api()
+      let validParams = []
+      let missingParams = []
+      for (let param of supportedRequests[requestName]) {
+        if (request.body[requestName][param]) {
+          validParams.push(request.body[requestName][param])
+        } else {
+          missingParams.push(param)
+        }
+      }
+      if (missingParams.length > 0) {
+        responseFn(ResponseMessage.Error(request, new Error(`Missing required parameter(s): ${missingParams}`)))
+      } else {
+        Promise.resolve(api[requestName](...validParams)).then((result) => {
+          responseFn(ResponseMessage.Success(request, result))
```

Comment:
One thing that may cause issues here is that the Arethusa instance is being controlled by the React application's state. But if the Arethusa instance is _also_ being controlled by an external application, the state of the Arethusa instance becomes out of sync with React's internal state.

For example, in the [demo page](https://alpheios-project.github.io/arethusa-demo-client/), the "Credits and more information" link goes to chunk [1](https://alpheios-project.github.io/treebank-template/on-the-murder-of-eratosthenes-1-50/1). If the user goes to ref `2` with the `Goto Ref` button, then the Arethusa instance shows chunk 2, but the "Credits and more information" link is still to chunk 1 (because the React application still thinks that the chunk is 1).

This seems like a really hard problem and I'm not sure how to fix it. I think we would need the callback to change the state of the React app (by changing the path) and then having React manipulate the Arethusa instance.

Reply:
I've made the new branch `wip-communication-interface` which adds some scaffolding in the [src/components/EmbeddedTreebank/EmbeddedTreebank.js](https://github.com/perseids-publications/treebank-template/commit/01581038b51a23625119e2b8bd471fe263bfba42#diff-79885fabf6b4a3cce7222f01f2f450f2) file.

What I think could work is checking out that branch (or rebasing) then taking this file, moving it to `src/components/EmbeddedTreebank/TreebankService.js`, and importing it in `EmbeddedTreebank.js`. Then in `registerTreebankListener()` you would call `active()`.

The code could be pretty much the same but instead of `let api = widget.api()` it would use the `gotoChunk`, `nextChunk`, and `prevChunk` functions defined in `TreebankService`. It may be a little more complex because those would have to be passed to the constructor but otherwise mostly the same.